### PR TITLE
NAS-101090 / 11.3 / Expose cloning method for jails in jail plugin

### PR DESCRIPTION
**Pull request by sonicaj**
This commit exposes api in jail plugin allowing users to clone jails via jail plugin.

### CODE REVIEW

File: src/middlewared/middlewared/plugins/jail.py [hunk 1]
```diff
@@ -214,6 +214,51 @@ def create_job(self, job, options):
...
+        Dict(
+            'clone_jail',
+            Str('uuid', required=True, empty=False),
+            List('pkglist', default=[], items=[Str('pkg', empty=False)]),
+            Bool('thickjail', default=False),
+            List('props', default=[]),
+        )
+    )
+    @job(lock=lambda args: f'clone_jail:{args[-2]}')
+    def clone_jail(self, job, source_jail, options):
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: Just `clone` is OK, we are already in jail namespace.

We probably want `@item_method` decorator here.

File: src/middlewared/middlewared/plugins/jail.py [hunk 2]
```diff
@@ -214,6 +214,51 @@ def create_job(self, job, options):
...
+
+        job.set_progress(25, 'Initial validation complete.')
+
+        ioc.IOCage(jail=source_jail, skip_jails=True).create(
+            source_jail, options['props'], _uuid=options['uuid'], thickjail=options['thickjail'], clone=True
+        )
+
+        job.set_progress(100, 'Jail has been successfully cloned.')
+
+        return self.middleware.call_sync('jail.query', [['host_hostuuid', '=', options['uuid']]], {'get': True})
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: What is `id` in jail.query? and why we cant use _get_instance?

File: src/middlewared/middlewared/plugins/jail.py [hunk 3]
```diff
@@ -214,6 +214,52 @@ def create_job(self, job, options):
...
+
+        job.set_progress(25, 'Initial validation complete.')
+
+        ioc.IOCage(jail=source_jail, skip_jails=True).create(
+            source_jail, options['props'], _uuid=options['uuid'], thickjail=options['thickjail'], clone=True
+        )
+
+        job.set_progress(100, 'Jail has been successfully cloned.')
+
+        return self._get_instance(options['uuid'])
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: Please test this, I think it wont work

File: src/middlewared/middlewared/plugins/jail.py [hunk 4]
```diff
@@ -214,6 +214,52 @@ def create_job(self, job, options):
...
+    )
+    @job(lock=lambda args: f'clone_jail:{args[-2]}')
+    def clone(self, job, source_jail, options):
+        verrors = ValidationErrors()
+        try:
+            self.check_jail_existence(source_jail, skip=False)
+        except CallError:
+            verrors.add(
+                'source_jail',
+                f'{source_jail} does not exist.'
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: While we are here, errno.ENOENT

File: src/middlewared/middlewared/plugins/jail.py [hunk 5]
```diff
@@ -214,6 +214,52 @@ def create_job(self, job, options):
...
+            )
+        else:
+            try:
+                self.check_jail_existence(options['uuid'], skip=False)
+            except CallError:
+                pass
+            else:
+                verrors.add(
+                    'clone_jail.uuid',
+                    f'Jail with "{options["uuid"]}" uuid already exists.'
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/jail.py]: errno.EEXIST

File: src/middlewared/middlewared/plugins/jail.py (hunk 3 above)
Comment (sonicaj) [reply src/middlewared/middlewared/plugins/jail.py]: it doesn't, already updated :)

**Pull request merged**